utilities/transactions/write_unprepared_transaction_test.cc

```diff
@@ -45,155 +48,244 @@ TEST_P(WriteUnpreparedTransactionTest, ReadYourOwnWrite) {
+
+      Transaction* txn;
+      TransactionOptions txn_options;
+      // batch_size of 1 causes writes to DB for every marker.
+      txn_options.max_write_batch_size = 1;
+      ReadOptions read_options;
+
+      for (uint32_t i = 0; i < NUM_ITERS; i++) {
+        std::set<std::string> owned_keys(&keys[id * NUM_KEYS],
+                                         &keys[(id + 1) * NUM_KEYS]);
+        // Add unowned keys to make the workload more interesting, but this
+        // increases row lock contention, so just do it sometimes.
+        if (rnd.OneIn(2)) {
+          owned_keys.insert(keys[rnd.Uniform(NUM_KEYS * NUM_THREADS)]);
+        }
 
-  iter->SeekToLast();
-  verify_state(iter, "b", "v4");
+        txn = db->BeginTransaction(write_options, txn_options);
+        txn->SetName(ToString(id));
+        txn->SetSnapshot();
+        if (a >= RO_SNAPSHOT) {
+          read_options.snapshot = txn->GetSnapshot();
+          ASSERT_TRUE(read_options.snapshot != nullptr);
+        }
 
-  iter->Prev();
-  verify_state(iter, "a", "v3");
+        uint64_t buf[2];
+        buf[0] = id;
 
-  roptions.snapshot = snapshot6;
-  wup_txn->unprep_seqs_[snapshot6->GetSequenceNumber() + 1] =
-      snapshot8->GetSequenceNumber() - snapshot6->GetSequenceNumber();
-  delete iter;
-  iter = txn->GetIterator(roptions);
+        // When scanning through the database, make sure that all unprepared
+        // keys have value >= snapshot and all other keys have value < snapshot.
+        int64_t snapshot_num = counter.fetch_add(1);
 
-  iter->SeekForPrev("b");
-  verify_state(iter, "b", "v8");
+        Status s;
+        for (const auto& key : owned_keys) {
+          buf[1] = counter.fetch_add(1);
+          s = txn->Put(key, Slice((const char*)buf, sizeof(buf)));
+          if (!s.ok()) {
+            break;
+          }
+          if (a == REFRESH_SNAPSHOT) {
+            txn->SetSnapshot();
+            read_options.snapshot = db->GetSnapshot();
+            snapshot_num = counter.fetch_add(1);
+          }
+        }
 
-  iter->Prev();
-  verify_state(iter, "a", "v7");
+        // Failure is possible due to snapshot validation. In this case,
+        // rollback and move onto next iteration.
+        if (!s.ok()) {
+          ASSERT_TRUE(s.IsBusy());
+          ASSERT_OK(txn->Rollback());
+          delete txn;
+          continue;
+        }
 
-  iter->SeekToLast();
-  verify_state(iter, "b", "v8");
+        auto verify_key = [&owned_keys, &a, &id, &snapshot_num](
+                              const std::string& key,
+                              const std::string& value) {
+          if (owned_keys.count(key) > 0) {
+            ASSERT_EQ(value.size(), 16);
+            ASSERT_EQ(((int64_t*)value.c_str())[0], id);
+            if (a == REFRESH_SNAPSHOT) {
+              // Since snapshot is refreshed after the last unprepared write,
+              // snapshot_num should be greater.
+              ASSERT_LT(((int64_t*)value.c_str())[1], snapshot_num);
```

Comment:
I am confused what is the we are verifying in this test. The act of the counter in this test and it being mixed with snapshot_num makes it more confusing.

Reply:
The counter is just a way of verifying that the snapshot is only seeing what it should be seeing.

In theory, we had access to the seqno of all the keys we read via Get() or iterator, then we can simply just compare it with the seqno of the snapshot we have.

Since I didn't see an easy way of getting the seqno of keys, I'm just keeping track of order with my own counter.

I can add more comments about what each check is doing, but I'm thinking of something like this:
```
if (owned_keys.count(key) > 0) {
  ASSERT_EQ(value.size(), 16);
  ASSERT_EQ(((int64_t*)value.c_str())[0], id); // Since this key is part of owned_keys, then this key must be unprepared by this transaction identified by 'id'
  if (a == REFRESH_SNAPSHOT) {
    ASSERT_LT(((int64_t*)value.c_str())[1], snapshot_num); // If refresh snapshot is true, then we refresh snapshot after every Put(), meaning that the current snapshot in snapshot_num must be greater than the "seqno" of any keys that we have written
  } else {
    ASSERT_GT(((int64_t*)value.c_str())[1], snapshot_num); // If refresh snapshot is not on, then we simply take a snapshot at the beginning of the transaction, meaning all writes must come _after_ snapshot_num
  }
} else if (a >= RO_SNAPSHOT) {
  ASSERT_EQ(value.size(), 16);
  ASSERT_LT(((int64_t*)value.c_str())[1], snapshot_num); // if this is not an unprepared key, just assert that the key "seqno" is smaller than the snapshot seqno.
}
```